### PR TITLE
Zachmu/query save feature

### PR DESCRIPTION
**Pull request by zachmu**
Implemented query saving via new dolt_query_catalog table, created by dolt sql -q -s.

Also:
* Unified business logic for validating table names in every code path where a table is created
* Separated out read-only, read-write, and alterable SQL tables
* Refactored deeply nested error handling logic in mv and cp commands
* Refactored import command to separate validation from execution logic
* Added many tests

### CODE REVIEW

File: bats/sql.bats [hunk 1]
```diff
@@ -48,6 +48,23 @@ teardown() {
     [ "${#lines[@]}" -eq 8 ]
 }
 
+@test "save query" {
```
Comment (timsehn) [inline bats/sql.bats]: Maybe this should be it's own bats test file?

File: go/libraries/doltcore/sqle/tables.go [hunk 2]
```diff
@@ -137,17 +147,17 @@ func (t *DoltTable) flushBatchedEdits(ctx context.Context) error {
...
-func (t *DoltTable) Replacer(ctx *sql.Context) sql.RowReplacer {
+func (t *WritableDoltTable) Replacer(ctx *sql.Context) sql.RowReplacer {
 	return t.getTableEditor()
 }
 
 // Updater implements sql.UpdatableTable
-func (t *DoltTable) Updater(ctx *sql.Context) sql.RowUpdater {
+func (t *WritableDoltTable) Updater(ctx *sql.Context) sql.RowUpdater {
 	return t.getTableEditor()
 }
```
Comment (bheni) [inline go/libraries/doltcore/sqle/tables.go]: These methods should immediately follow the struct declaration.

File: go/libraries/doltcore/sqle/tables.go [hunk 3]
```diff
@@ -291,7 +301,7 @@ func (t *DoltTable) DropColumn(ctx *sql.Context, columnName string) error {
...
 
 // ModifyColumn implements sql.AlterableTable
-func (t *DoltTable) ModifyColumn(ctx *sql.Context, columnName string, column *sql.Column, order *sql.ColumnOrder) error {
+func (t *AlterableDoltTable) ModifyColumn(ctx *sql.Context, columnName string, column *sql.Column, order *sql.ColumnOrder) error {
 	table, _, err := t.db.Root().GetTable(ctx, t.name)
 	if err != nil {
 		return err
```
Comment (bheni) [inline go/libraries/doltcore/sqle/tables.go]: These methods should immediately follow the struct declaration

File: go/cmd/dolt/commands/tblcmds/table.go [hunk 4]
```diff
@@ -25,3 +27,14 @@ var Commands = cli.NewSubCommandHandler("table", "Commands for copying, renaming
 	MvCmd{},
 	CpCmd{},
 })
+
+// Validates the given table name for creation as a user table.
```
Comment (bheni) [inline go/cmd/dolt/commands/tblcmds/table.go]: Comments for exported items should begin with the exact name of the item

File: go/libraries/doltcore/doltdb/table.go [hunk 5]
```diff
@@ -36,8 +39,25 @@ const (
 
 	// TableNameRegexStr is the regular expression that valid tables must match.
 	TableNameRegexStr = `^[a-zA-Z]{1}$|^[a-zA-Z]+[-_0-9a-zA-Z]*[0-9a-zA-Z]+$`
+
+	// We reserve all tables that begin with dolt_ for system use.
+	DoltNamespace = "dolt_"
```
Comment (bheni) [inline go/libraries/doltcore/doltdb/table.go]: Comments for exported items should begin with the exact name of the item

File: go/libraries/doltcore/sql/sqltestutil/schema.go [hunk 6]
```diff
@@ -72,17 +72,25 @@ func NewResultSetRow(colVals ...types.Value) row.Row {
 // Creates a new row with the values given, using ascending tag numbers starting at 0.
 // Uses the first value as the primary key.
 func NewRow(colVals ...types.Value) row.Row {
+	return NewRowWithPks(colVals[0:1], colVals[1:]...)
+}
+
+// Creates a new row with the values given, using ascending tag numbers starting at 0.
+func NewRowWithPks(pkColVals []types.Value, nonPkVals ...types.Value) row.Row {
```
Comment (bheni) [inline go/libraries/doltcore/sql/sqltestutil/schema.go]: Comments for exported items should begin with the exact name of the item

File: go/libraries/doltcore/sql/sqltestutil/schema.go [hunk 7]
```diff
@@ -72,17 +72,25 @@ func NewResultSetRow(colVals ...types.Value) row.Row {
 // Creates a new row with the values given, using ascending tag numbers starting at 0.
 // Uses the first value as the primary key.
 func NewRow(colVals ...types.Value) row.Row {
```
Comment (bheni) [inline go/libraries/doltcore/sql/sqltestutil/schema.go]: Comments for exported items should begin with the exact name of the item

File: go/libraries/doltcore/sqle/database.go [hunk 8]
```diff
@@ -141,34 +146,48 @@ func (db *Database) GetTableInsensitive(ctx context.Context, tblName string) (sq
...
+
+	db.tables[exactName] = toReturn
+	return toReturn, true, nil
 }
 
+// Returns the names of all user tables. System tables in user space (e.g. dolt_docs, dolt_query_catalog) are filtered
+// out. This method is used for queries that examine the schema of the database, e.g. show tables. Table name resolution
+// in queries is handled by GetTableInsensitive. Use GetAllTableNames for an unfiltered list of all tables in user
+// space.
 func (db *Database) GetTableNames(ctx context.Context) ([]string, error) {
```
Comment (bheni) [inline go/libraries/doltcore/sqle/database.go]: Comments for exported items should begin with the exact name of the item

File: go/libraries/doltcore/sqle/database.go [hunk 9]
```diff
@@ -263,8 +291,14 @@ func (db *Database) RenameTable(ctx *sql.Context, oldName, newName string) error
 // Flushes the current batch of outstanding changes and returns any errors.
 func (db *Database) Flush(ctx context.Context) error {
```
Comment (bheni) [inline go/libraries/doltcore/sqle/database.go]: Comments for exported items should begin with the exact name of the item

File: go/libraries/doltcore/sqle/database.go [hunk 10]
```diff
@@ -302,15 +336,15 @@ func (db *Database) CreateView(ctx *sql.Context, name string, definition string)
 // dolt database. Returns sql.ErrNonExistingView if the view did not
 // exist.
 func (db *Database) DropView(ctx *sql.Context, name string) error {
```
Comment (bheni) [inline go/libraries/doltcore/sqle/database.go]: Comments for exported items should begin with the exact name of the item

File: go/libraries/doltcore/sqle/database.go [hunk 11]
```diff
@@ -335,19 +370,19 @@ func (db *Database) DropView(ctx *sql.Context, name string) error {
 // schema fragments if they don't parse, or if registries within the
 // `catalog` return errors.
 func RegisterSchemaFragments(ctx *sql.Context, catalog *sql.Catalog, db *Database) error {
```
Comment (bheni) [inline go/libraries/doltcore/sqle/database.go]: Comments for exported items should begin with the exact name of the item

File: go/libraries/doltcore/sqle/schema_table.go [hunk 12]
```diff
@@ -16,50 +16,50 @@ package sqle
...
 		// For example, the SELECT statement for a CREATE VIEW.
-		{Name: "fragment", Type: sql.Text, Source: LogTableName, PrimaryKey: false},
+		{Name: "fragment", Type: sql.Text, Source: doltdb.SchemasTableName, PrimaryKey: false},
 	}
 }
 
 // Returns the `dolt_schemas` table in `db`, creating it if it does not already exist.
-func GetOrCreateDoltSchemasTable(ctx *sql.Context, db *Database) (*DoltTable, error) {
-	tbl, found, err := db.GetTableInsensitive(ctx, SchemasTableName)
+func GetOrCreateDoltSchemasTable(ctx *sql.Context, db *Database) (*WritableDoltTable, error) {
```
Comment (bheni) [inline go/libraries/doltcore/sqle/schema_table.go]: Comments for exported items should begin with the exact name of the item

File: bats/sql.bats (hunk 1 above)
Comment (bheni) [reply bats/sql.bats]: would be nice to see a conflict test in bats also.

File: go/cmd/dolt/commands/sql.go [hunk 13]
```diff
@@ -105,6 +108,9 @@ func (cmd SqlCmd) createArgParser() *argparser.ArgParser {
 	ap := argparser.NewArgParser()
 	ap.SupportsString(queryFlag, "q", "SQL query to run", "Runs a single query and exits")
 	ap.SupportsString(formatFlag, "r", "Result output format", "How to format result output. Valid values are tabular, csv. Defaults to tabular. ")
+	ap.SupportsFlag(saveFlag, "s", "Used with --query, save the query to the query catalog. Saved queries can be examined in the dolt_query_catalog system table.")
```
[the remark this answers is not in the document]
Comment (zachmu) [reply go/cmd/dolt/commands/sql.go]: I settled on making --save the name flag, basically making name required.

File: bats/sql.bats (hunk 1 above)
Comment (zachmu) [reply bats/sql.bats]: Done.

**Pull request merged**